pkg/http/api_directmessaging.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/valyala/fasthttp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	diagConsts "github.com/dapr/dapr/pkg/diagnostics/consts"
+	"github.com/dapr/dapr/pkg/http/endpoints"
+	"github.com/dapr/dapr/pkg/messages"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	"github.com/dapr/dapr/pkg/resiliency"
+	"github.com/dapr/dapr/pkg/resiliency/breaker"
+)
+
+// directMessagingSpanData is the data passed by the onDirectMessage endpoint to the tracing middleware
+type directMessagingSpanData struct {
+	// Target app ID
+	AppID string
+	// Method invoked
+	Method string
+}
+
+func appendDirectMessagingSpanAttributes(r *http.Request, m map[string]string) {
+	endpointData, _ := r.Context().Value(endpoints.EndpointCtxKey{}).(*endpoints.EndpointCtxData)
+	if endpointData != nil && endpointData.SpanData != nil {
+		spanData, _ := endpointData.SpanData.(*directMessagingSpanData)
+		if spanData != nil {
+			m[diagConsts.GrpcServiceSpanAttributeKey] = "ServiceInvocation"
+			m[diagConsts.NetPeerNameSpanAttributeKey] = spanData.AppID
+			m[diagConsts.DaprAPISpanNameInternal] = "CallLocal/" + spanData.AppID + "/" + spanData.Method
+		}
+	}
+}
+
+func (a *api) constructDirectMessagingEndpoints() []endpoints.Endpoint {
+	return []endpoints.Endpoint{
+		{
+			// No method is defined here to match any method
+			Methods: []string{},
+			Route:   "invoke/*",
+			// This is the fallback route for when no other method is matched by the router
+			Version: apiVersionV1,
+			Group: &endpoints.EndpointGroup{
+				Name:                 endpoints.EndpointGroupServiceInvocation,
+				Version:              endpoints.EndpointGroupVersion1,
+				AppendSpanAttributes: appendDirectMessagingSpanAttributes,
+			},
+			Handler: a.onDirectMessage,
+			Settings: endpoints.EndpointSettings{
+				Name:                  "InvokeService",
+				IsFallback:            true,
+				KeepWildcardUnescaped: true,
+			},
+		},
+	}
+}
+
+// onDirectMessage is the handler for HTTP service invocation
+func (a *api) onDirectMessage(w http.ResponseWriter, r *http.Request) {
```

Comment:
This method, and many more in this file, has been moved from `api.go` without changes, to keep things more organized